Export Calendar events to ICS files - Issue #50

AmilaDias opened the pull request:
Implemented ICS export functionality as per issue #50 .

iCalExport Class- 
Created to handle export
iCalExport takes a Event object and will output ICS files
Current location is hard coded into project folder, would prefer to send
this to a server or directly into Google Calendar or iCalendar via APIs
(would require login/separate server, so new issue/enhancement?).

MenuController.Java - 
Added call for iCalExport, thus when calendar is selected from menu the
system will export the calendar entries into ICS files. A button could
be implemented for export.
iExport.iCalSetup establishes the file folder for the export, this
method could be directed towards a server or API integration

Supporting Libraries
Biweekly (ICS export) - https://github.com/mangstadt/biweekly.git
- Uses BSD 2-Clause license
- Vinnie (Supporting library for Biweekly)

AmilaDias commented:
Travis seems to fail due to it not detecting the new class I added (ICalExport.java) and the libraries I imported (Biweekly and Vinnie). Is this something I can repair, or is it going to be a problem until the pull request is merged?

rsanchez-wsu commented on src/Model/ICalExport.java:
```diff
@@ -0,0 +1,133 @@
+		
+		//Set duration within event
+		Duration duration = new Duration.Builder().hours(hours).minutes(minutes).build();
+		calEvent.setDuration(duration);	
```
This trailing white space needs to be removed.

rsanchez-wsu commented on src/Model/ICalExport.java:
```diff
@@ -0,0 +1,133 @@
+
+		//Set start date
+		calEvent.setDateStart(eStart);
+				
```
This trailing white space needs to be removed.

rsanchez-wsu commented on src/Model/ICalExport.java:
```diff
@@ -0,0 +1,133 @@
+				
+		//Add new event to ICS File
+		ical.addEvent(calEvent);
+		
```
This trailing white space needs to be removed.

rsanchez-wsu commented on src/Model/ICalExport.java:
```diff
@@ -0,0 +1,133 @@
+		//Create ICS File
+		createIcs(ical, file);
+	}
+	
```
This trailing white space needs to be removed.

rsanchez-wsu commented on src/Model/ICalExport.java:
```diff
@@ -0,0 +1,133 @@
+			newDir.mkdirs();
+		}
+	}
+	
```
This trailing white space needs to be removed.

rsanchez-wsu commented on src/Model/ICalExport.java:
```diff
@@ -0,0 +1,133 @@
+		}
+		
+	}
+	
```
This trailing white space needs to be removed.

rsanchez-wsu commented on src/Model/ICalExport.java:
```diff
@@ -0,0 +1,133 @@
+	private int minutes;
+	private int counter;
+	
+	public void createExportEvent(Event event, int counter){
```
A space is required before the opening brace.

rsanchez-wsu commented on src/Model/ICalExport.java:
```diff
@@ -0,0 +1,133 @@
+		createIcs(ical, file);
+	}
+	
+	public void iCalSetup(){
```
A space is required before the opening brace.

rsanchez-wsu commented on src/Model/ICalExport.java:
```diff
@@ -0,0 +1,133 @@
+		}
+	}
+	
+	private void createIcs(ICalendar iCal, File file){
```
A space is required before the opening brace.

rsanchez-wsu commented on src/Model/ICalExport.java:
```diff
@@ -0,0 +1,133 @@
+		
+	}
+	
+	private String createFileName(int counter){
```
A space is required before the opening brace.

rsanchez-wsu commented on src/Model/ICalExport.java:
```diff
@@ -128,6 +118,5 @@ public void setTitle(String title) {
 	 * @param description the description to set
 	 */
 	public void setDescription(MultilineString description) {
-		this.description = description;
```
The removal of this line does not seem correct.  The setter now does nothing.

rsanchez-wsu commented on src/Model/ICalExport.java:
```diff
@@ -32,36 +32,30 @@
 		seteStart(event.getDate());
 		setTitle(event.getName());
 		setDescription(event.getDetails());
-		this.counter = counter;
```
I do not understand why this line was removed.  It seems like the value of counter is no longer saved.

rsanchez-wsu commented on src/Controller/DataController.java:
```diff
@@ -427,9 +427,7 @@ static private HubFile processNewHubFile(NodeList nList) throws Exception
             ConsoleIO.setConsoleMessage("Attempting to import " + Integer.toString(assetList.size())
                     + " items to VersionControl...", true);
             ConsoleIO.setConsoleMessage("Starting with " + VersionControlEntity.libraryReport() + " entries");
-
```
This should likely have remained as a blank line to maintain readability.

AmilaDias replied:
This has been rectified in current commits, white space here is fixed and located at line 417 in DataController

rsanchez-wsu commented on src/Controller/DataController.java:
```diff
@@ -427,9 +427,7 @@ static private HubFile processNewHubFile(NodeList nList) throws Exception
             ConsoleIO.setConsoleMessage("Starting with " + VersionControlEntity.libraryReport() + " entries");
-
             ArrayList<Event> calendarItems = new ArrayList<>();
-            
```
This should likely have remained as a blank line to maintain readability.

rsanchez-wsu commented on src/Controller/MenuController.java:
```diff
@@ -444,15 +444,12 @@ public void loadCalendar() {
 		// =================
 		// Populate Agenda:
 		ArrayList<Event> calendar = MainController.getSPC().getPlanner().getCurrentStudyProfile().getCalendar();
-		
```
This should likely have remained as a blank line to maintain readability.

rsanchez-wsu commented on src/Controller/MenuController.java:
```diff
@@ -444,15 +444,12 @@ public void loadCalendar() {
         ICalExport iExport = new ICalExport();
         //Preparation of export directory
         iExport.iCalSetup();
-		
```
One of these should likely have remained as a blank line to maintain readability.

rsanchez-wsu commented on src/Model/iCalExport.java:
```diff
@@ -0,0 +1,134 @@
+		VEvent calEvent = new VEvent();
+		
+		Summary summary = calEvent.setSummary(title);
+		summary.setLanguage("en-us");
```
Rather than setting the language like this, the program should first check the environment variable `LANG`.  If that variable is set, then the language specified there should be attempted first.  If that does not work, then it is acceptable to default to something like this.  Please also add a note to the issue on the application preferences framework that documents a need for a language preference in the application.  When that is implemented, then we would have the user preference in the application as another possibility.

rsanchez-wsu commented on src/Model/iCalExport.java:
```diff
@@ -0,0 +1,134 @@
+		Summary summary = calEvent.setSummary(title);
+		summary.setLanguage("en-us");
+		
+		//Default hard coded duration
```
The default duration should not be set from hard coded numbers.  Constants (i.e., final static, whether public or private) should be declared and then used here.  Additionally, the default should likely be a duration of zero.  Your if-else below does not handle Activity or Deadline types, so a default of zero would be more sensible for something like a deadline that has no duration.

rsanchez-wsu commented on src/Model/iCalExport.java:
```diff
@@ -0,0 +1,134 @@
+		minutes = 0;
+		
+		//If duration exists within event, pull duration
+		if (event instanceof TimetableEvent) {
```
I really hate this code.  However, I understand that you have no other way to accomplish this.  Please add a `TODO` note here that indicates that once #88 has been fixed that this can be discarded replaced with a single call to `event.getDuration()`.

rsanchez-wsu commented on src/Model/iCalExport.java:
```diff
@@ -0,0 +1,134 @@
+	
+	public void iCalSetup(){
+		File newDir = new File("calendarExport");
+		if(newDir.exists()){
```
A space is required before the ( and before the {.

rsanchez-wsu commented on src/Model/iCalExport.java:
```diff
@@ -0,0 +1,134 @@
+		}
+	}
+	
+	private void createIcs(ICalendar iCal, File file){
```
A space is required before the {.

rsanchez-wsu commented on src/Model/iCalExport.java:
```diff
@@ -0,0 +1,134 @@
+	}
+	
+	private void createIcs(ICalendar iCal, File file){
+		try{
```
A space is required before the {.

rsanchez-wsu commented on src/Model/iCalExport.java:
```diff
@@ -0,0 +1,134 @@
+	private void createIcs(ICalendar iCal, File file){
+		try{
+			Biweekly.write(iCal).go(file);
+		} catch (IOException e){
```
A space is required before the {.

AmilaDias commented:
To answer the question above @rsanchez-wsu, I did not change StudyPlanner.dat, so I'm not exactly sure how that modification happened. If anything, I have tried to implement my code in the least destructive method possible, keeping the original project code in tact as much as possible.

AmilaDias commented:
I believe I've taken care of all the requisite issues. I have reverted .classpath and there git doesn't show a difference between the new SamplePlanner.dat file and the file kept in my local master branch. I've also fixed (to the best of my knowledge) all of the trailing white space/checkstyle issues that were present.

rsanchez-wsu commented on .classpath:
```diff
@@ -2,7 +2,32 @@
 <classpath>
 	<classpathentry kind="src" path="src"/>
 	<classpathentry kind="src" path="Test"/>
+<<<<<<< HEAD
```
It looks like you left the conflicted section in here.  Please remove the conflict markers and the libraries that were added back in.

rsanchez-wsu commented on src/Controller/DataController.java:
```diff
@@ -414,9 +414,9 @@ private static HubFile processNewHubFile(NodeList nList) throws Exception
                     + " items to VersionControl...", true);
             ConsoleIO.setConsoleMessage("Starting with " + VersionControlEntity.libraryReport() + " entries");
-
+            
```
Trailing whitespace.

rsanchez-wsu commented on src/Controller/DataController.java:
```diff
@@ -414,9 +414,9 @@ private static HubFile processNewHubFile(NodeList nList) throws Exception
+            
             ArrayList<Event> calendarItems = new ArrayList<>();
-
+            
```
Trailing whitespace.

rsanchez-wsu commented on src/Model/iCalExport.java:
```diff
@@ -0,0 +1,138 @@
+		String str = "calendarExport" + counter + ".ics";
+		return str;
+	}
+	
```
Trailing whitespace.

rsanchez-wsu commented on src/Model/iCalExport.java:
```diff
@@ -0,0 +1,138 @@
+ *
+ */
+public class iCalExport {
+	
```
Trailing whitespace.

rsanchez-wsu commented on src/Model/iCalExport.java:
```diff
@@ -0,0 +1,138 @@
+	private int hours;
+	private int minutes;
+	private int counter;
+	
```
Trailing whitespace.

AmilaDias commented:
Addressed all changes requested.
